lib/connectors/mongodb/connector.rb

```diff
@@ -56,30 +70,166 @@ def initialize(configuration: {})
+
+        function, args = setup_aggregate(filter_config) if aggregate_present?(filter_config)
+
+        [function, args]
+      end
+
+      def check_filter_config!(rules, config)
+        return unless filtering_present?(rules, config)
+
+        check_find_and_aggregate_present!(config)
+
+        check_find_and_aggregate_missing!(config)
+      end
+
+      def check_find_and_aggregate_missing!(config)
+        allowed_keys = Utility::Strings.format_string_array(ALLOWED_TOP_LEVEL_FILTER_KEYS, default: ' ')
+        keys_present = Utility::Strings.format_string_array(config.keys, default: ' ')
+
+        find_and_aggregate_missing = !find_present?(config) && !aggregate_present?(config)
+        both_missing_msg = "Only one of #{allowed_keys} is allowed in the filtering object. Keys present: '#{keys_present}'."
+
+        raise Utility::InvalidFilterConfigError.new(both_missing_msg) if find_and_aggregate_missing
+      end
+
+      def check_find_and_aggregate_present!(config)
+        find_and_aggregate_present = find_present?(config) && aggregate_present?(config)
+        both_present_msg = '\'find\' and \'aggregate\' functions cannot be used at the same time for MongoDB. Please drop one from the configuration.'
+
+        raise Utility::InvalidFilterConfigError.new(both_present_msg) if find_and_aggregate_present
+      end
+
+      def call_db_function_on_collection(function, args, &on_doc_serialization)
         with_client do |client|
           # We do paging using skip().limit() here to make Ruby recycle the memory for each page pulled from the server after it's not needed any more.
           # This gives us more control on the usage of the memory (we can adjust PAGE_SIZE constant for that to decrease max memory consumption).
           # It's done due to the fact that usage of .find.each leads to memory leaks or overuse of memory - the whole result set seems to stay in memory
           # during the sync. Sometimes (not 100% sure) it even leads to a real leak, when the memory for these objects is never recycled.
-          cursor = client[@collection].find
+          cursor = client[@collection].send(function, *args)
           skip = 0
 
+          options = args[1]
+          found_overall = 0
+
+          # if no overall limit is specified by filtering use -1 to not break ingestion, when no overall limit is specified (found_overall is only increased,
+          # thus can never reach -1)
+          overall_limit = -1
+
+          unless options.nil?
+            # there could be a skip parameter defined for filtering
+            skip = Utility::Commons.return_if_present(options[:skip], 0)
+            # there could be a limit parameter defined for filtering -> used for an overall limit (not a page limit, which was introduced for memory optimization)
+            overall_limit = Utility::Commons.return_if_present(options[:limit], -1)
+          end
```

Comment:
I'm not sure exposing/allowing skip/limit brings value. While `limit` might make sense for some customer, `skip` does not seem to have any benefit but can provide weird experience if user copy/pasted the query incorrectly.

Reply:
I don't want to try to set artificial constraints here. The implementation doc suggested we let the user provide whatever options they want so that they can tune how they fetch data. `Skip` could let the user skip an initial set of records that they don't care to index. `limit` lets them set the batch size as it makes sense for their performance.